http_101.rst

```diff
@@ -1,20 +1,369 @@
+	                                "Server: myserver"\
+        	                        "Content-Length: 30"\
+                	                ""\
+                        	        "<html><b>Not Found</b></html>" | tee >(cat - >&2);
+	        fi
+        	unset request method uri version
+	        exit 0
+	}
+
+	## MAIN ##
+
+	read -r request
+	read -r method uri version <<<"$request"
+	[ -n "$method" ] && [ -n "$uri" ] && [ -n "$version" ] || echo "HTTP/1.1 400 Bad Request"
+	echo $request >&2
+	while read -r header
+	do
+        	header=${header%%$'\r'}
+	        echo $header >&2
+	        [ -z "$header" ] && { respond; break; }
+	done
+
+
+Stdout output acts as the response to the client.
+Output to file descriptor 2 (``>&2``) is written to the server terminal.
+This only covers reprentation of a web page, or an error page in case of a missing file.
+
+To launch this with netcat compiled with DGAPING_SECURITY_HOLE option:
+
+.. code-block:: console
+
+	$ while true; do netcat -lp 1234 -e ./httpd.sh; done
+
+Or when ``-e`` is not available:
+
+.. code-block:: console
+
+	$ mkfifo /tmp/httpipe
+	$ while true; do cat /tmp/httpipe | ./httpd.sh | nc -l 192.168.0.1 1234 > /tmp/httpipe ; done
+
+This creates a pipe that we use to redirect data to and from the httpd script.
+Now we can test it with either of the above shown tools:
+
+.. code-block:: console
+
+	$ curl 192.168.0.1:1234/
+	<html>Hello there world!</html>
+
+	$ nc 192.168.0.1 1234
+	GET /notexisting.html HTTP/1.1
+
+	HTTP/1.1 404 Not Found
+	Date: Tue, 9 Jun 18:39:59 GMT
+	Server: myserver
+	Content-Length: 30
+
+	<html><b>Not Found</b></html>
+
+	$ telnet 192.168.0.1 1234
+	Trying 192.168.0.1...
+	Connected to 192.168.0.1.
+	Escape character is '^]'.
+	GET /index.html HTTP/1.1
+
+	HTTP/1.1 200 OK
+	Date: Tue, 9 Jun 18:44:53 GMT
+	Server: myserver
+	Content-Length: 32
+
+	<html>Hello there world!</html>
+	Connection closed by foreign host.
+
 
 Apache, nginx
 =============
 
+Apache and Nginx are the two most used open-source webservers on the internet.
+Though they share several features, architecturally they differ significally.
+
+Apache relies on a process-based model, either with a single or multiple threads per process that handle a connection, whereas Nginx operates event-driven with multiple single-threaded processes that all handle connections asynchronously.
+Nginx' event-driven aproach requires less resources, making it faster, especially for static content.
+For dynamic content however, Nginx needs to proxy requests to an external processor for execution as it has no programming language support.
+Therefore Nginx is often used as a front-end/proxy server in combination with other application servers, possibly Apache.
+
+Both web servers are extensible with modules (security, compression, proxy, ...).
+Where Apache is able to dynamically load and enable modules at runtime, Nginx need to have additional modules compiled in the core to have it available.
```

Comment:
Nginx need to have additional modules compiled in the core to have it available. 
=>
Nginx can only load modules that have been precompiled into the executable.